Adding CARAFE: Content-Aware ReAssembly of FEatures

jordanesikati opened the pull request:
Add a new module: CARAFE, an operator for feature upsampling in convolutional neural networks.

- Large field of view. 
- Content-aware handling
- Lightweight and fast computing.

Paper: https://arxiv.org/pdf/1905.02188.pdf

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
This PR adds new model configurations and introduces the CARAFE upsampling module, along with improvements to attention modules in the Ultralytics YOLO ecosystem. 🚀

### 📊 Key Changes
- **New Model Configurations:**  
  - Added `yolov8-carafe.yaml` for object detection using the CARAFE upsampling module.
  - Added `yolov8-cls-vgg16.yaml` for image classification using a VGG16 backbone.
- **CARAFE Module:**  
  - Implemented the CARAFE (Content-Aware ReAssembly of FEatures) upsampling layer for improved feature upsampling.
  - Integrated CARAFE into the model parsing and module imports.
- **Attention Module Enhancements:**  
  - Refactored Channel Attention, Spatial Attention, and CBAM (Convolutional Block Attention Module) for better clarity and performance.
- **General Code Improvements:**  
  - Updated module imports and model parsing logic to support new modules.

### 🎯 Purpose & Impact
- **Enhanced Model Flexibility:**  
  - Users can now experiment with CARAFE-based upsampling for potentially better detection accuracy, especially for small objects. 🔍
  - The new VGG16-based classification model offers more backbone choices for image classification tasks.
- **Better Feature Representation:**  
  - Improved attention modules (CBAM, Channel, and Spatial Attention) can help models focus on important features, possibly boosting performance.
- **Broader Research & Customization:**  
  - These additions make it easier for researchers and developers to try advanced architectures and compare results within the Ultralytics framework.

Overall, this update brings more options and improved components for building and experimenting with YOLO models! 🛠️✨

jordanesikati committed:
Create yolov8-carafe.yaml

jordanesikati committed:
Update block.py

jordanesikati committed:
Update __init__.py

jordanesikati committed:
Update tasks.py

Add CARAFE

jordanesikati committed:
Merge pull request #1 from jordanesikati/jordanesikati-patch-1

Create yolov8-carafe.yaml

jordanesikati committed:
Merge pull request #2 from jordanesikati/jordanesikati-patch-2

Update block.py

jordanesikati committed:
Merge pull request #3 from jordanesikati/jordanesikati-patch-3

Update __init__.py

jordanesikati committed:
Merge pull request #4 from jordanesikati/jordanesikati-patch-4

Update tasks.py

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

jordanesikati committed:
Update block.py

jordanesikati committed:
Merge pull request #5 from jordanesikati/jordanesikati-patch-5

Update block.py

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

jordanesikati commented:
I have read the CLA Document and I sign the CLA

jordanesikati committed:
Update block.py

jordanesikati committed:
Update block.py

jordanesikati committed:
Merge pull request #6 from jordanesikati/jordanesikati-patch-6

Update block.py

glenn-jocher commented:
@jordanesikati ah very cool, thanks for the PR! Have you tried training this branch on COCO for before and after results?

jordanesikati commented:
> @jordanesikati ah very cool, thanks for the PR! Have you tried training this branch on COCO for before and after results?

Due to hardware constraints, I trained this branch on a custom dataset from which I am doing some research, and it showed quite good results based on mAP0.5 and mAP0.5-0.95. I am not able to provide you with the dataset because of privacy concerns (biomedical-related dataset). In addition, I used the same technique with YOLOv7 on another custom dataset (AppleScabFDs), resulting in a 2.5% improvement based on mAP0.5, and 1.8% based on mAP0.5-0.95.

jordanesikati committed:
Create yolov8-cls-vgg16.yaml

jordanesikati committed:
Merge pull request #7 from jordanesikati/jordanesikati-patch-7

Create yolov8-cls-vgg16.yaml

jordanesikati committed:
Merge branch 'main' into main

glenn-jocher commented:
Thank you for your contribution, @jordanesikati! It's great to hear about the positive results you've seen using this branch on your custom datasets. While different datasets will always yield a variety of results, your experience suggests that this feature could have considerable potential.

It would be fantastic if collaborators with the necessary resources could train this branch on COCO dataset for a more standardized comparison. Meanwhile, your provided metrics based on your custom dataset provide valuable insight for the rest of the community. 

Please make sure to continue sharing your findings and thank you again for your contribution to the repo. Appreciation belongs to the entire Ultralytics team and our brilliant community contributors such as yourself. Your privacy concerns are completely understood and respected. Please keep us updated on any further advancements or findings!

jordanesikati commented:
> 

@glenn-jocher Sure!

glenn-jocher commented:
Thank you, @jordanesikati! We're excited to hear more about your future results. Your contributions are valuable to our work here at Ultralytics and the overall YOLOv8 community. Continue testing and pushing the limits; and remember, every little improvement can lead to significant advancements in the performance of the model. Happy experimenting!

jordanesikati commented:
> Thank you, @jordanesikati! We're excited to hear more about your future results. Your contributions are valuable to our work here at Ultralytics and the overall YOLOv8 community. Continue testing and pushing the limits; and remember, every little improvement can lead to significant advancements in the performance of the model. Happy experimenting!

@glenn-jocher Thank you.

glenn-jocher committed:
Merge branch 'main' into main

jordanesikati commented:
> @jordanesikati I'd like to train on COCO, can you provide some steps to reproduce? Do I just point to the new YAMLs, i.e. `yolov8s-carafe.yaml`?

@glenn-jocher Yes, of course

# Ultralytics YOLO 🚀, AGPL-3.0 license
# YOLOv8 object detection model with P3-P5 outputs. For Usage examples see https://docs.ultralytics.com/tasks/detect

# Parameters
nc: 80  # number of classes
scales: # model compound scaling constants, i.e. 'model=yolov8n.yaml' will call yolov8.yaml with scale 'n'
  # [depth, width, max_channels]
  n: [0.33, 0.25, 1024]  # YOLOv8n summary: 225 layers,  3157200 parameters,  3157184 gradients,   8.9 GFLOPs
  s: [0.33, 0.50, 1024]  # YOLOv8s summary: 225 layers, 11166560 parameters, 11166544 gradients,  28.8 GFLOPs
  m: [0.67, 0.75, 768]   # YOLOv8m summary: 295 layers, 25902640 parameters, 25902624 gradients,  79.3 GFLOPs
  l: [1.00, 1.00, 512]   # YOLOv8l summary: 365 layers, 43691520 parameters, 43691504 gradients, 165.7 GFLOPs
  x: [1.00, 1.25, 512]   # YOLOv8x summary: 365 layers, 68229648 parameters, 68229632 gradients, 258.5 GFLOPs

# YOLOv8.0n backbone
backbone:
  # [from, repeats, module, args]
  - [-1, 1, Conv, [64, 3, 2]]  # 0-P1/2
  - [-1, 1, Conv, [128, 3, 2]]  # 1-P2/4
  - [-1, 3, C2f, [128, True]]
  - [-1, 1, Conv, [256, 3, 2]]  # 3-P3/8
  - [-1, 6, C2f, [256, True]]
  - [-1, 1, Conv, [512, 3, 2]]  # 5-P4/16
  - [-1, 6, C2f, [512, True]]
  - [-1, 1, Conv, [1024, 3, 2]]  # 7-P5/32
  - [-1, 3, C2f, [1024, True]]
  - [-1, 1, SPPF, [1024, 5]]  # 9

# YOLOv8.0n head
head:
#  - [-1, 1, nn.Upsample, [None, 2, 'nearest']]
  - [-1, 1, CARAFE, [512, 3, 2]]
  - [[-1, 6], 1, Concat, [1]]  # cat backbone P4
  - [-1, 3, C2f, [512]]  # 12

#  - [-1, 1, nn.Upsample, [None, 2, 'nearest']]
  - [-1, 1, CARAFE, [256, 3, 2]]
  - [[-1, 4], 1, Concat, [1]]  # cat backbone P3
  - [-1, 3, C2f, [256]]  # 15 (P3/8-small)

  - [-1, 1, Conv, [256, 3, 2]]
  - [[-1, 12], 1, Concat, [1]]  # cat head P4
  - [-1, 3, C2f, [512]]  # 18 (P4/16-medium)

  - [-1, 1, Conv, [512, 3, 2]]
  - [[-1, 9], 1, Concat, [1]]  # cat head P5
  - [-1, 3, C2f, [1024]]  # 21 (P5/32-large)

  - [[15, 18, 21], 1, Detect, [nc]]  # Detect(P3, P4, P5)

glenn-jocher commented:
@jordanesikati got it! 

I started training with `profile=True`, which will export the models to ONNX and TensorRT to profile them after the first epoch is complete, but this ran into errors. 

It seems that the models are not exportable due to the torch unfold operation not existing in ONNX. Do you know if there's any method to export to ONNX? 

Without support for all of our export formats it will be difficult to use these models more broadly. See https://docs.ultralytics.com/modes/export for examples.

glenn-jocher committed:
Merge branch 'main' into main

jordanesikati commented:
> @jordanesikati got it!
> 
> I started training with `profile=True`, which will export the models to ONNX and TensorRT to profile them after the first epoch is complete, but this ran into errors.
> 
> It seems that the models are not exportable due to the torch unfold operation not existing in ONNX. Do you know if there's any method to export to ONNX?
> 
> Without support for all of our export formats it will be difficult to use these models more broadly. See https://docs.ultralytics.com/modes/export for examples.

@glenn-jocher I have not tried to export into ONNX or other formats yet. I will figure it out later.

glenn-jocher committed:
Merge branch 'main' into main

glenn-jocher committed:
Merge branch 'main' into main

glenn-jocher committed:
Merge branch 'main' into main

glenn-jocher committed:
Merge branch 'main' into main

glenn-jocher committed:
Merge branch 'main' into main

glenn-jocher committed:
Merge branch 'main' into main

glenn-jocher committed:
Merge branch 'main' into main

jordanesikati committed:
Merge branch 'ultralytics:main' into main

jordanesikati committed:
Merge branch 'ultralytics:main' into main

jordanesikati committed:
Merge branch 'ultralytics:main' into main

jordanesikati commented:
> > @jordanesikati got it!
> > I started training with `profile=True`, which will export the models to ONNX and TensorRT to profile them after the first epoch is complete, but this ran into errors.
> > It seems that the models are not exportable due to the torch unfold operation not existing in ONNX. Do you know if there's any method to export to ONNX?
> > Without support for all of our export formats it will be difficult to use these models more broadly. See https://docs.ultralytics.com/modes/export for examples.
> 
> @glenn-jocher I have not tried to export into ONNX or other formats yet. I will figure it out later.

* This is the configuration used to train coco using yolov8n-carafe with "profile=True" enabled
[dummy training COCO 1 epoch]
![coco-training](https://github.com/ultralytics/ultralytics/assets/119162141/8066e762-9156-44a9-b0ab-adfad37f2391)

* Verbose
![one_epoch_done](https://github.com/ultralytics/ultralytics/assets/119162141/113b1dbe-2d53-4162-bc95-97e4d42b11d3)

![coco-verbose](https://github.com/ultralytics/ultralytics/assets/119162141/4014f85a-8756-4c99-a9da-67aa8204fa04)

* ONNX export
![onnx](https://github.com/ul
![onnx_export_success](https://github.com/ultralytics/ultralytics/assets/119162141/cd694676-fb80-4a80-bb91-6386c74b51e5)
tralytics/ultralytics/assets/119162141/59884273-62a7-4cd4-b4e6-7aae2924625b)

jordanesikati commented:
Hi @glenn-jocher, I will find enough time to work on the training for 300+ epochs.

glenn-jocher commented:
@jordanesikati That sounds great! We appreciate your dedication and look forward to seeing the results. Training the models for 300+ epochs is indeed a significant task. Don't hesitate to report back on the progress you're making or any interesting findings you have during the process. We're always here to discuss and provide any necessary support. Happy training!

jordanesikati committed:
Merge branch 'ultralytics:main' into main

jordanesikati committed:
Fixed rtdet custom training

Remediate "ValueError: matrix contains invalid numeric entries" while training RT-DETR on a custom dataset.

jordanesikati committed:
Merge pull request #8 from jordanesikati/rtdetr-custom-training

Fixed rtdet custom training

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

jordanesikati committed:
Merge branch 'ultralytics:main' into main

jordanesikati committed:
Update ops.py

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

msabrimas commented:
> > @jordanesikatiSaya ingin berlatih tentang COCO, dapatkah Anda memberikan beberapa langkah untuk mereproduksi? Apakah saya hanya menunjuk ke YAML baru saja `yolov8s-carafe.yaml`?
> 
> @glenn-jocherYa, tentu saja
> 
> https://github.com/jordanesikati/ultralytics/blob/main/ultralytics/cfg/models/v8/yolov8-carafe.yaml

carafe module not found, even i using your repo, what should i do?

glenn-jocher commented:
@msabrimas sure, here's the possible reason and solution.

The `carafe` module may not be available directly in PyTorch but is used in the YOLOv8-CARAFE model, which is why you're seeing the error. It probably means that the CARAFE layer has not been properly installed or imported in your set-up. 

To solve this, try the following steps:

1. Ensure PyTorch is correctly installed in your environment.
2. You may need to download and compile the CARAFE layer manually in order to use it in the model. This often involves running a setup or install file included with the CARAFE source code.
3. After installing the CARAFE layer, ensure it's correctly imported in your Python scripts where the YOLOv8-CARAFE model is being constructed.

Remember the exact steps to follow could be different based on your system and setup. So, please read and follow any instructions provided with the CARAFE source code carefully.

jordanesikati committed:
Merge branch 'ultralytics:main' into main

jordanesikati committed:
Refactoring CBAM 

Reimplementation of  CBAM (Convolutional Block Attention Module) in conv.py

Paper: https://arxiv.org/abs/1807.06521 

Default kernel_size=7

Default reduction ratio r=16

Definition of  Shared MLP

jordanesikati committed:
Merge pull request #9 from jordanesikati/fixing-cbam-1

Refactoring CBAM

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

jordanesikati committed:
Fixing Key Error: 'CBAM'

jordanesikati committed:
Merge pull request #10 from jordanesikati/fixing-cbam-2

Fixing Key Error: 'CBAM'

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

jordanesikati committed:
Reinsert CARAFE

Deletion while inserting CBAM in tasks.py

jordanesikati committed:
Merge pull request #11 from jordanesikati/reinsert-carafe

Reinsert CARAFE

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

jordanesikati committed:
Refactoring CBAM

In Spatial Attention Module 
Correction

jordanesikati committed:
Merge pull request #12 from jordanesikati/fixing-cbam-3

Refactoring CBAM

triangel-nuo commented:
Hello, based on the CARAFE you submitted, I modified the corresponding file exactly as you did, but the following error occurred:
Traceback (most recent call last):
  File "train.py", line 5, in <module>
    model = YOLO('cfg/models/v8/yolov8-carafe.yaml')  # 加载预训练模型（推荐用于训练）
  File "/home/hp/anaconda3/envs/yolov8/lib/python3.8/site-packages/ultralytics/engine/model.py", line 92, in __init__
    self._new(model, task)
  File "/home/hp/anaconda3/envs/yolov8/lib/python3.8/site-packages/ultralytics/engine/model.py", line 128, in _new
    self.model = (model or self._smart_load('model'))(cfg_dict, verbose=verbose and RANK == -1)  # build model
  File "/home/hp/anaconda3/envs/yolov8/lib/python3.8/site-packages/ultralytics/nn/tasks.py", line 232, in __init__
    self.model, self.save = parse_model(deepcopy(self.yaml), ch=ch, verbose=verbose)  # model, savelist
  File "/home/hp/anaconda3/envs/yolov8/lib/python3.8/site-packages/ultralytics/nn/tasks.py", line 678, in parse_model
    m = getattr(torch.nn, m[3:]) if 'nn.' in m else globals()[m]  # get module
KeyError: 'CARAFE'

I would like to ask what is the cause of this and how to solve it. It would be my honor to receive your answer. Thank you very much.

He-Yingchao commented:
> > > @jordanesikati got it!
> > > I started training with `profile=True`, which will export the models to ONNX and TensorRT to profile them after the first epoch is complete, but this ran into errors.
> > > It seems that the models are not exportable due to the torch unfold operation not existing in ONNX. Do you know if there's any method to export to ONNX?
> > > Without support for all of our export formats it will be difficult to use these models more broadly. See https://docs.ultralytics.com/modes/export for examples.
> > 
> > 
> > @glenn-jocher I have not tried to export into ONNX or other formats yet. I will figure it out later.
> 
> Hi @glenn-jocher
> 
> * This is the configuration used to train coco using yolov8n-carafe with "profile=True" enabled
>   [dummy training COCO 1 epoch]
>   ![coco-training](https://private-user-images.githubusercontent.com/119162141/267986189-8066e762-9156-44a9-b0ab-adfad37f2391.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDg0ODUxOTgsIm5iZiI6MTcwODQ4NDg5OCwicGF0aCI6Ii8xMTkxNjIxNDEvMjY3OTg2MTg5LTgwNjZlNzYyLTkxNTYtNDRhOS1iMGFiLWFkZmFkMzdmMjM5MS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMjIxJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDIyMVQwMzA4MThaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT05NGUzMDk1M2EyMzg1YjljY2ZiZDRiMWZkNDgyYmQyZGVlMzU1ZDkyMzE1YjJkNjlkYWFhMzQ0NTJiYjJmNjNkJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.hCQvgPCszdKCSsss-JNmsI0ZBbC6qBoh4Brn6WFf3io)
> * Verbose 1
>   ![one_epoch_done](https://private-user-images.githubusercontent.com/119162141/267999943-113b1dbe-2d53-4162-bc95-97e4d42b11d3.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDg0ODUxOTgsIm5iZiI6MTcwODQ4NDg5OCwicGF0aCI6Ii8xMTkxNjIxNDEvMjY3OTk5OTQzLTExM2IxZGJlLTJkNTMtNDE2Mi1iYzk1LTk3ZTRkNDJiMTFkMy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMjIxJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDIyMVQwMzA4MThaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0yMGNmNzljZGMxNTliZThiYmUwNGQ4ZDFlOTkyMTBjNjlkOWYzN2I5ODM0MzRlMzVkYzU0OGViNjVhOGYyMWY3JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.grl-lG7WxE9Vh_vvP0d0JBTP3Va2sjFndl_gt0pBNow)
> * Verbose 2
>   ![coco-verbose](https://private-user-images.githubusercontent.com/119162141/267998123-4014f85a-8756-4c99-a9da-67aa8204fa04.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDg0ODUxOTgsIm5iZiI6MTcwODQ4NDg5OCwicGF0aCI6Ii8xMTkxNjIxNDEvMjY3OTk4MTIzLTQwMTRmODVhLTg3NTYtNGM5OS1hOWRhLTY3YWE4MjA0ZmEwNC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMjIxJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDIyMVQwMzA4MThaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1iNGY5MjI4NGJlNDg5ZWVkOWFhMjM3NzhjNzc1ZjAzZTE4ZGNmMzNiMzRiMDA0NmYwYmZhMDhmYjMzNzMwZDZjJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.XPW15ALJO2VsVV9itxgUBnW_wXktcKXzIxZodvXjFOY)
> * ONNX export (Script)
>   ![onnx](https://private-user-images.githubusercontent.com/119162141/268004840-63002529-89b3-4ba1-a9d2-fe74c32b36a4.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDg0ODUxOTgsIm5iZiI6MTcwODQ4NDg5OCwicGF0aCI6Ii8xMTkxNjIxNDEvMjY4MDA0ODQwLTYzMDAyNTI5LTg5YjMtNGJhMS1hOWQyLWZlNzRjMzJiMzZhNC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMjIxJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDIyMVQwMzA4MThaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1lZmZhNzRlMTcwNWY0YTZkYzJhYzI0ZDY2MGJjOWM4NjhmMjFlZWYyZjdhMzdhNmUzOTFhMmY2YTExNTQwNWJhJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.y5H4XnVp-WgLuVZk88AgAgXliZmcvbpWzAjSc_NDxTc)
> * ONNX export (Success)
>   ![onnx_export_success](https://private-user-images.githubusercontent.com/119162141/268002965-cd694676-fb80-4a80-bb91-6386c74b51e5.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDg0ODUxOTgsIm5iZiI6MTcwODQ4NDg5OCwicGF0aCI6Ii8xMTkxNjIxNDEvMjY4MDAyOTY1LWNkNjk0Njc2LWZiODAtNGE4MC1iYjkxLTYzODZjNzRiNTFlNS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMjIxJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDIyMVQwMzA4MThaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT00NmMzZWQ2OWI0ODRlYzY5YWQ4YjU0ZjM0ODQyODg1YmYyY2U3YmZmZTZiMmQ5MjMxMDAzN2ZmZjVmZGRkMjRlJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.vuuFL9eidxEPO_ufXsiWMmjNRItwVlOnFkYe8SjZqx0)
>   tralytics/ultralytics/assets/119162141/59884273-62a7-4cd4-b4e6-7aae2924625b)

I used the carafe module in my yaml file, and the trained pt file failed to export onnx. Later, when I retrained, I added profile=true, but I met the same problem as you. Have you solved it?

glenn-jocher commented:
Hi @He-Yingchao 👋,

It looks like the CARAFE module isn't recognized during the export process. This is likely because custom modules like CARAFE need to be supported with specialized code for ONNX export. 

For now, you might consider using standard PyTorch operations that have direct ONNX support for your upsampling needs, or look into ONNX's custom op support to implement CARAFE. If you're set on using CARAFE, keep an eye on updates or consider contributing to the repo with an ONNX-compatible version of the module. 😊

Hope this helps!
